### PR TITLE
ARROW-476: Add binary integration test fixture, add Java support

### PR DESCRIPTION
**Pull request by wesm**
@julienledem could you review my Java changes? Thanks

### CODE REVIEW

File: java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java [hunk 1]
```diff
@@ -199,6 +200,9 @@ private void setValueFromParser(ValueVector valueVector, int i) throws IOExcepti
     case FLOAT8:
       ((Float8Vector)valueVector).getMutator().set(i, parser.readValueAs(Double.class));
       break;
+    case VARBINARY:
+      ((VarBinaryVector)valueVector).getMutator().setSafe(i, parser.readValueAs(String.class).getBytes(UTF_8));
+      break;
```
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java]: a binary value may not be valid UTF-8
do we want to encode in hexadecimal for this?

Comment (wesm) [inline java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java]: Right -- the integration tests are only generating ASCII strings so this works for now. Hexadecimal would be better. I can do this in Python / C++ if you have some Java code to convert to `byte[]` from a hex string

Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java]: you can use commons-codec's Hex class:
you can add this dependency to the pom:
```
    <dependency>
      <groupId>commons-codec</groupId>
      <artifactId>commons-codec</artifactId>
      <version>1.10</version>
      <scope>compile</scope>
    </dependency>
```

If you don't want to add a dependency you can also convert one byte at a time:
String.format("%02x", byteValue);

**Comment by wesm** [comment]
@julienledem ok, this should be good to go

**Comment by wesm** [comment]
Just waiting on the OS X build here https://travis-ci.org/wesm/arrow/builds/199737937

File: java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java [hunk 2]
```diff
@@ -164,6 +167,14 @@ private void readVector(Field field, FieldVector vector) throws JsonParseExcepti
     readToken(END_OBJECT);
   }
 
+  private byte[] decodeHexSafe(String hexString) throws IOException {
+    try {
+      return Hex.decodeHex(hexString.toCharArray());
+    } catch (DecoderException e) {
+      throw new IOException("Unable to decode hex string: " + hexString);
```
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java]: Exceptions should be chained:
throw new IOException("Unable to decode hex string: " + hexString, e);

**Review by julienledem — APPROVED** [review]
Just one minor comment, otherwise LGTM

File: cpp/src/arrow/ipc/json-internal.cc [hunk 3]
```diff
@@ -852,6 +883,8 @@ class JsonArrayReader {
     const auto& json_data_arr = json_data->value.GetArray();
 
     DCHECK_EQ(static_cast<int32_t>(json_data_arr.Size()), length);
+
+    std::vector<uint8_t> byte_buffer;
```
Comment (xhochy) [inline cpp/src/arrow/ipc/json-internal.cc]: This should be a `PoolBuffer`

**Review by xhochy — CHANGES_REQUESTED** [review]
We should avoid using `std::vector` for data.

**Comment by wesm** [comment]
Done

**Review by xhochy — APPROVED** [review]
+1, LGTM

**Comment by wesm** [comment]
Green build: https://travis-ci.org/wesm/arrow/builds/199986996

**Pull request closed without merging**